services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -247,6 +290,68 @@ private void updatePortToHost(PortState portState) {
 
     }
 
+    private void updatePortStateAsync(PortState portState, PortState oldPortState, AsyncExecutor executor,
+                                      Stack<PortStateRollback> rollbacks) throws Exception {
+        //Update mac_address
+        String macAddress = portState.getMacAddress();
+        String oldMacAddress = oldPortState.getMacAddress();
+        if (macAddress != null && !oldMacAddress.equals(macAddress)) {
+            MacManagerProxy macManagerProxy = new MacManagerProxy(rollbacks);
+            executor.runAsync(macManagerProxy::releaseMacAddress, oldPortState);
+            executor.runAsync(macManagerProxy::allocateFixedMacAddress, portState);
+            oldPortState.setMacAddress(macAddress);
+        }
+
+        //Update device_owner and device_id
+        String deviceOwnerNew = portState.getDeviceOwner();
+        String deviceIdNew = portState.getDeviceId();
+        String deviceIdOld = oldPortState.getDeviceId();
+        String tenantId = oldPortState.getTenantId();
+
+        if (deviceOwnerNew != null && deviceIdNew != null && !deviceIdNew.equals(deviceIdOld)) {
+            if (DeviceOwner.ROUTER.getOwner().equals(deviceOwnerNew)) {
+                verifyRouter(deviceIdNew, tenantId);
+            }
+        }
+
+        //Update fixed_ips
+        List<PortState.FixedIp> fixedIps = portState.getFixedIps();
+        IpManagerProxy ipManagerProxy = new IpManagerProxy(rollbacks, portState.getProjectId());
+
+        if (fixedIps != null) {
+            List<PortState.FixedIp> oldFixedIps = oldPortState.getFixedIps();
+
+            List<PortState.FixedIp> addFixedIps = fixedIpsCompare(fixedIps, oldFixedIps);
+            List<PortState.FixedIp> delFixedIps = fixedIpsCompare(oldFixedIps, fixedIps);
+
+            if (delFixedIps.size() > 0) {
+                executor.runAsync(ipManagerProxy::releaseIpAddressBulk, delFixedIps);
+            }
+
+            if (addFixedIps.size() > 0) {
+                executor.runAsync(ipManagerProxy::allocateFixedIpAddress, addFixedIps);
+            }
```

Comment:
Same there. Allocate before release.

Also, could you give a simple comment or example to explain addFixedIps and delFixedIps to increase code readability?

Reply:
Same replacement API might be needed for PIP manager as well.

Reply:
Tracked by Issue https://github.com/futurewei-cloud/alcor/issues/229.